runtime.cljs

```diff
@@ -33,50 +37,76 @@
-                                                                (.split "\n")
-                                                                (.slice 1))})})
+                  :body (js/JSON.stringify runtime-error)})
         (.then (fn [{:keys [status]
                      :as res}]
-                 (assert (= status 200)
-                         (str "Unexpected " url "response: " (js/JSON.stringify res)))
+                 ;; We treat all errors from the API as an unrecoverable error. This is
+                 ;; because the API returns 4xx errors for responses that are too long. In
+                 ;; that case, we simply log the output and fail.
+                 (assert (successful? status) (str "Error from Runtime API\n"
+                                                   (-> {:url url
+                                                        :response res
+                                                        :error runtime-error}
+                                                       (clj->js)
+                                                       (js/JSON.stringify nil 2))))
                  res)))))
 
-(defonce state (atom nil))
-
 (defn start []
-  (-> (request {:url (str runtime-path "/invocation/next")})
-      (.then (fn [{:keys                                                                       [status body]
-                   {aws-request-id                      "lambda-runtime-aws-request-id"
-                    lambda-runtime-invoked-function-arn "lambda-runtime-invoked-function-arn"} :headers
-                   :as                                                                         response}]
-               (let [context {:aws-request-id aws-request-id
-                              :lambda-runtime-invoked-function-arn lambda-runtime-invoked-function-arn}]
-                 (swap! state assoc :context context)
-                 (assert (= status 200) (str "Unexpected /invocation/next response: " (pr-str response)))
-                 {:event   (-> (js/JSON.parse body)
-                               js->clj)
-                  :context context})))
-      (.then handle)
-      (.then (fn [response]
-               (let [{:keys [aws-request-id]} (:context @state)]
-                 (request {:url    (str runtime-path "/invocation/" aws-request-id "/response")
-                           :method :post
-                           :headers {"Content-Type" "application/json"}
-                           :body   (-> (clj->js response)
-                                       js/JSON.stringify)}))))
-      (.then (fn [{:keys [status]
-                   :as response}]
-               (assert (= status 202) (str "Unexpected /invocation/response response:" (pr-str response)))))
-      (.catch (fn [err]
-                (post-error {:error err
-                             :context (:context @state)})))
-      (.then start)))
+  (let [state (atom nil)
```

Comment:
I'm not sure if let'ing the state is a good idea. Won't we leak an atom each time an invocation of `start` is done? 

I agree that `def`ing an atom is not the best way, but maybe a better way would be to pass the reference to the atom each time we call `start` and only make a new one on the initial invocation? 

In any case it should be another PR, since it's not related to error handling as far as I see.

Reply:
Yeah well, yes if you want to share state between warm invocations.

I might have misread it but the I thought the info in that atom was per-invocation only and strictly only necessary for the downstream promises.

Let me know the actual intent, of course I can change it back 😁

Reply:
Yes, it's correct, it's only intended for downstream promises, and having it in a `let` would seem to be the best choice, except I'm not sure what happens when we recur (call `start` again in a `then`) - does the old atom get garbage collected? If so, your solution is good. If not, we should either keep the old way with `def`'ing an atom, or simply pass on the atom in the `start` invocation.

In both cases, let's keep this refactor for a new PR, in order to keep one feature per PR :-)